### PR TITLE
Clean-up billing API tests

### PR DESCRIPTION
**Pull request by hodgestar**
Currently there's a lot of duplicated code and a lot of things being tested in individual tests.

### CODE REVIEW

**Comment by hodgestar** [comment]
This needs to be done before #881 because that really needs to add more transaction creation API tests and those are a bit of a mess right now.

**Comment by hodgestar** [comment]
Ready for review.

**Comment by hodgestar** [comment]
Would really like to split these into separate tests, but that requires solving the issue of how to wrap the entire connection in a transaction and roll it back at the end of the test (essentially duplicating what Django does in its tests). That's probably best left for a separate PR -- things are already a lot cleaner.

**Comment by hodgestar** [comment]
Alternatively we could wrap the tests in `django_db(transaction=True)` at the cost of making the tests **really** slow (because the database has to be deleted and recreated before each test).

**Comment by hodgestar** [comment]
An acceptable hack might be to clear all the relevant tables at the end of each test?

**Comment by smn** [comment]
I'm not sure Django creates & deletes the db for every test, I thought it just ran it in a single transaction which it then rolled back.

**Comment by smn** [comment]
but that may be _entirely_ irrelevant to this PR so ignore me for now.

**Comment by hodgestar** [comment]
Django doesn't by default, but one can ask it too. If we rely on Django's
transaction rollback that doesn't clean up the database because the changes
made by the billing api happen in their own connection.

**Comment by smn** [comment]
thanks :cake:

File: go/billing/tests/test_api.py [hunk 1]
```diff
@@ -25,218 +25,211 @@
...
-
-        self.assertEqual(response.responseCode, 200)
-        new_user = json.loads(response.value(), cls=JSONDecoder)
-        self.assertTrue('id' in new_user)
-
-        # Fetch the new user
-        url = 'users/%s' % (new_user.get('id'))
-        response = yield self.web.get(url)
+        http_method = getattr(self.web, method)
+        response = yield http_method(path, headers=headers, **kw)
```
Comment (justinvdm) [inline go/billing/tests/test_api.py:48]: Would there ever be a case where there were other headers besides `{'content-type': 'application/json'}`?

File: go/billing/tests/test_api.py [hunk 2]
```diff
@@ -25,218 +25,211 @@
...
-        url = 'accounts/%s' % (new_account.get('account_number'))
-        response = yield self.web.get(url)
-        self.assertEqual(response.responseCode, 200)
-        account = json.loads(response.value(), cls=JSONDecoder)
-        self.assertEqual(account.get('account_number'),
-                         new_account.get('account_number'))
-        self.assertTrue(account.get('credit_balance', 0) == 100)
+        # test retrieving all users
+        user_list = yield self.get_api_user_list()
+        self.assertEqual(user_list, [new_user])
```
Comment (justinvdm) [inline go/billing/tests/test_api.py:184]: Testing retrieving all users feels like it should be in a separate test. A bit of a bikeshed though, so your call.

**Comment by justinvdm** [comment]
It seem seem like most of the test cases are testing a few things inside a single test. Any reason to keep it that way? Sorry, I know this is a little bikesheddy. Happy for it to stay as is if it needs to.

File: go/billing/tests/test_api.py (hunk 1 above)
Comment (hodgestar) [inline go/billing/tests/test_api.py:48]: Not with the API as it stands (it's meant to be all JSON over HTTP).

Comment (justinvdm) [inline go/billing/tests/test_api.py:48]: Ok cool.

File: go/billing/tests/test_api.py (hunk 2 above)
Comment (hodgestar) [inline go/billing/tests/test_api.py:184]: Totally agree. I wanted to do that initially but it runs into problems with
how the database is managed (see discussion I had on this PR this with
@smn). Would like to solve those, but not in this PR (which unfortunately
is thus limited to a smaller set of cleaning up).

Comment (justinvdm) [inline go/billing/tests/test_api.py:184]: Ah, ok, np then. Its at least quite a bit cleaner and readable now, so that can wait.

Comment (hodgestar) [inline go/billing/tests/test_api.py:184]: I created #886 for splitting up the tests.

**Comment by justinvdm** [comment]
:+1: I tried check if we are missing anything important from the assertions in the original tests, but can't seem to find anything. Much cleaner now :)

**Comment by justinvdm** [comment]
:+1:

**Pull request merged**